presto-hive/src/main/java/io/prestosql/plugin/hive/security/SqlStandardAccessControlMetadata.java

```diff
@@ -84,6 +87,62 @@ private static void checkRoleIsNotReserved(String role)
         return ImmutableSet.copyOf(metastore.listRoles());
     }
 
+    @Override
+    public Set<RoleGrant> listAllRoleGrants(ConnectorSession session, Optional<Set<String>> roles, Optional<Set<String>> grantees, OptionalLong limit)
+    {
+        /*
+         * The Hive Metadata interface provides us with two ways to retrieve role grants:
+         * 1. principals by single role
+         * 2. roles by single principal
+         *
+         * We check the passed predicates to decide which path is most optimal.
+         */
+        if (roles.isPresent() && grantees.isPresent()) {
+            if (roles.get().size() <= grantees.get().size() * 2) {
+                // 2x because for the grantees we do not know the grantee type (ROLE or USER)
+                grantees = Optional.empty();
+            }
+            else {
+                roles = Optional.empty();
+            }
+        }
+
+        ImmutableSet.Builder<RoleGrant> builder = ImmutableSet.builder();
+        long count = 0;
+        if (grantees.isPresent()) {
+            TOP:
+            for (String grantee : grantees.get()) {
+                for (PrincipalType type : new PrincipalType[]{USER, ROLE}) {
+                    for (RoleGrant grant : metastore.listRoleGrants(new HivePrincipal(type, grantee))) {
+                        if (limit.isPresent() && count >= limit.getAsLong()) {
+                            break TOP;
+                        }
+                        // Filter out the "public" role since it is not explicitly granted in Hive.
+                        if ("public".equals(grant.getRoleName())) {
+                            continue;
+                        }
+                        count++;
+                        builder.add(grant);
+                    }
+                }
+            }
+            return builder.build();
+        }
+
+        // if no roles or grantees were passed get a list of roles from the metastore
+        for (String role : roles.orElseGet(() -> metastore.listRoles())) {
```

Comment:
please extract one method for collecting output for `grantees` and one method for `roles`

Reply:
Done.